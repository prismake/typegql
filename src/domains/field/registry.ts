import { DeepWeakMap } from '~/services/utils';

export interface AllRegisteredFields {
  [fieldName: string]: FieldInnerConfig;
}

export const fieldsRegistry = new DeepWeakMap<
  Function,
  FieldInnerConfig,
  AllRegisteredFields
>();

export interface FieldInnerConfig {
  name: string;
  property: string;
  description?: string;
  isNullable?: boolean;
  isGetter?: boolean;
  type?: any;
}

export interface AllQueryFields {
  [fieldName: string]: FieldInnerConfig;
}

export const queryFieldsRegistry = new DeepWeakMap<
  Function,
  FieldInnerConfig,
  AllQueryFields
>();
