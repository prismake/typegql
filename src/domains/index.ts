export {
  ObjectType,
  compileObjectType,
  objectTypeRegistry,
} from './objectType';
export {
  InterfaceType,
  compileInterfaceType,
  interfaceTypeRegistry,
} from './interfaceType';
export {
  InputObjectType,
  compileInputObjectType,
  inputObjectTypeRegistry,
} from './inputObjectType';
export { Field } from './field';
export { InputField } from './inputField';
export { Arg } from './arg';
export { Inject, Context, Source, Info } from './inject';
export { registerEnum, enumsRegistry } from './enum';
export { Union, unionRegistry } from './union';
export { Before, After } from './hooks';
export {
  SchemaRoot,
  Schema,
  schemaRootsRegistry,
  compileSchema,
  Query,
  Mutation,
} from './schema';
