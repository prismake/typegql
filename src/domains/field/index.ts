import { fieldsRegistry, FieldInnerConfig } from './registry';

export {
  FieldInnerConfig,
  fieldsRegistry,
  queryFieldsRegistry,
} from './registry';
export { compileAllFields, compileFieldConfig } from './compiler';
export { isQueryField } from './services';
export { FieldError } from './error';

export interface FieldOptions {
  description?: string;
  type?: any;
  name?: string;
  isNullable?: boolean;
}

export function Field(options?: FieldOptions): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    const finalConfig: FieldInnerConfig = {
      property: fieldName,
      name: fieldName,
      isNullable: true,
      ...options,
    };

    const descriptor = Object.getOwnPropertyDescriptor(
      targetInstance,
      fieldName,
    );

    fieldsRegistry.set(targetInstance.constructor, fieldName, {
      isGetter: descriptor && !!descriptor.get,
      ...finalConfig,
    });
  };
}
