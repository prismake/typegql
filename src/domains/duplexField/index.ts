import { fieldsRegistry, FieldInnerConfig } from '../field/registry';
import {
  inputFieldsRegistry,
  FieldInputInnerConfig,
} from '../inputField/registry';

export { FieldError } from './error';

export function DuplexField(
  options?: FieldInnerConfig &
    FieldInputInnerConfig & {
      inputNullable?: boolean;
    },
): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    let isNullable = true;
    let inputNullable = true;
    if (options) {
      if (options.isNullable !== undefined) {
        isNullable = options.isNullable;
      }
      if (options.inputNullable !== undefined) {
        inputNullable = options.inputNullable;
      }
      delete options.isNullable;
      delete options.inputNullable;
    }

    const finalInputConfig: FieldInputInnerConfig = {
      property: fieldName,
      name: fieldName,
      isNullable: inputNullable,
      ...options,
    } as any;

    const finalConfig: FieldInnerConfig = {
      property: fieldName,
      name: fieldName,
      isNullable,
      ...options,
    } as any;
    fieldsRegistry.set(targetInstance.constructor, fieldName, finalConfig);

    inputFieldsRegistry.set(
      targetInstance.constructor,
      fieldName,
      finalInputConfig,
    );
  };
}
