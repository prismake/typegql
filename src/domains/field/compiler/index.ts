import { GraphQLFieldConfig, GraphQLFieldConfigMap } from 'graphql';
import { getClassWithAllParentClasses } from '~/services/utils/inheritance';
import { FieldError, fieldsRegistry } from '../index';

import { compileFieldResolver } from './resolver';
import { compileFieldArgs } from '~/domains/arg';
import {
  enhanceType,
  isRootFieldOnNonRootBase,
  resolveRegisteredOrInferedType,
  validateResolvedType,
} from './services';

import { validateNotInferableField } from './fieldType';

export function compileFieldConfig(
  target: Function,
  fieldName: string,
): GraphQLFieldConfig<any, any, any> {
  const { type, description, isNullable, isGetter } = fieldsRegistry.get(
    target,
    fieldName,
  );

  const args = compileFieldArgs(target, fieldName);
  const resolvedType = resolveRegisteredOrInferedType(target, fieldName, type);

  // if was not able to resolve type, try to show some helpful information about it
  if (!resolvedType && !validateNotInferableField(target, fieldName)) {
    return;
  }

  // show error about being not able to resolve field type
  if (!validateResolvedType(target, fieldName, resolvedType)) {
    validateNotInferableField(target, fieldName);
    return;
  }

  const finalType = enhanceType(resolvedType, isNullable);

  return {
    description,
    type: finalType,
    resolve: compileFieldResolver(target, fieldName, isGetter),
    args,
  };
}

function getAllFields(target: Function) {
  const fields = fieldsRegistry.getAll(target);
  const finalFieldsMap: GraphQLFieldConfigMap<any, any> = {};
  Object.keys(fields).forEach(fieldName => {
    if (isRootFieldOnNonRootBase(target, fieldName)) {
      throw new FieldError(
        target,
        fieldName,
        `Given field is root field (@Query or @Mutation) not registered inside @Schema type. `,
      );
    }

    const config = fieldsRegistry.get(target, fieldName);
    finalFieldsMap[config.name] = compileFieldConfig(target, fieldName);
  });
  return finalFieldsMap;
}

export function compileAllFields(target: Function) {
  const targetWithParents = getClassWithAllParentClasses(target);

  const finalFieldsMap: GraphQLFieldConfigMap<any, any> = {};

  targetWithParents.forEach(targetLevel => {
    Object.assign(finalFieldsMap, getAllFields(targetLevel));
  });
  return finalFieldsMap;
}
