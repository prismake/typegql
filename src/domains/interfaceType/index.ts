import { compileInterfaceTypeWithConfig } from './compiler';
import { interfaceTypeRegistry } from './registry';

export { compileInterfaceType } from './compiler';
export { InterfaceTypeError } from './error';
export { interfaceTypeRegistry } from './registry';

export interface InterfaceTypeOptions {
  name?: string;
  description?: string;
}

export function InterfaceType(options?: InterfaceTypeOptions): ClassDecorator {
  return (target: Function) => {
    const config = { name: target.name, ...options };
    const outputTypeCompiler = () =>
      compileInterfaceTypeWithConfig(target, config);
    interfaceTypeRegistry.set(target, outputTypeCompiler);
  };
}
