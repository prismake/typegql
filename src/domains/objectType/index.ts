import { compileObjectTypeWithConfig } from './compiler';
import { objectTypeRegistry } from './registry';

export { compileObjectType } from './compiler';
export { ObjectTypeError } from './error';
export { objectTypeRegistry } from './registry';

export interface ObjectTypeOptions {
  name?: string;
  description?: string;
}

export function ObjectType(options?: ObjectTypeOptions): ClassDecorator {
  return (target: Function) => {
    const config = { name: target.name, ...options };
    const outputTypeCompiler = () =>
      compileObjectTypeWithConfig(target, config);
    objectTypeRegistry.set(target, outputTypeCompiler);
  };
}
