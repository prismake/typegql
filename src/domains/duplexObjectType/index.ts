import { objectTypeRegistry } from '../objectType/registry';
import { compileObjectTypeWithConfig } from '../objectType/compiler';
import { compileInputObjectTypeWithConfig } from '../inputObjectType/compiler';
import { inputObjectTypeRegistry } from '../inputObjectType/registry';
export { ObjectTypeError } from './error';

export interface ObjectTypeOptions {
  name?: string;
  description?: string;
}

export function DuplexObjectType(options?: ObjectTypeOptions): ClassDecorator {
  return (target: Function) => {
    const inputConfig = { name: target.name + 'Input', ...options };
    const outputConfig = { name: target.name, ...options };

    const inputTypeCompiler = () =>
      compileInputObjectTypeWithConfig(target, inputConfig);

    inputObjectTypeRegistry.set(target, inputTypeCompiler);

    const outputTypeCompiler = () =>
      compileObjectTypeWithConfig(target, outputConfig);
    objectTypeRegistry.set(target, outputTypeCompiler);
  };
}
