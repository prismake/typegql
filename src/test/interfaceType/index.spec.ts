import { GraphQLInterfaceType } from 'graphql';
import { InterfaceType, compileInterfaceType } from '~/domains';
import { Field } from '~/domains/field';

describe('Type', () => {
  it('Throws when trying to compile type without @InterfaceType decorator', () => {
    expect(() =>
      compileInterfaceType(class Bar {}),
    ).toThrowErrorMatchingSnapshot();
  });

  it('Throws when @InterfaceType has no fields', () => {
    @InterfaceType()
    class NoFields {}

    @InterfaceType()
    class NoDeclaredFields {
      foo: string;
    }
    expect(() => compileInterfaceType(NoFields)).toThrowErrorMatchingSnapshot();
    expect(() =>
      compileInterfaceType(NoDeclaredFields),
    ).toThrowErrorMatchingSnapshot();
  });

  it('Compiles basic type with field', () => {
    @InterfaceType()
    class Foo {
      @Field() bar: string;
    }

    const compiled = compileInterfaceType(Foo);

    const fields = compiled.getFields();
    const barField = fields.bar;

    expect(compiled).toBeInstanceOf(GraphQLInterfaceType);

    expect(barField).toBeTruthy();
    expect(barField.name).toEqual('bar');
  });

  it('Sets proper options', () => {
    @InterfaceType({ description: 'Baz' })
    class Foo {
      @Field() bar: string;
    }

    const compiled = compileInterfaceType(Foo);

    expect(compiled.description).toEqual('Baz');
    expect(compiled.name).toEqual('Foo');

    @InterfaceType({ name: 'Baz' })
    class FooCustomName {
      @Field() bar: string;
    }

    const compiledCustomName = compileInterfaceType(FooCustomName);

    expect(compiledCustomName.name).toEqual('Baz');
  });

  it('Final type is compiled only once per class', () => {
    @InterfaceType()
    class Foo {
      @Field() bar: string;
    }

    const compiledA = compileInterfaceType(Foo);
    const compiledB = compileInterfaceType(Foo);

    expect(compiledA).toBe(compiledB);
  });
});
