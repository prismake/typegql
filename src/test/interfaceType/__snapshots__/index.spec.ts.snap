// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Type Throws when @InterfaceType has no fields 1`] = `"@InterfaceType 'NoFields': There are no fields inside this type."`;

exports[`Type Throws when @InterfaceType has no fields 2`] = `"@InterfaceType 'NoDeclaredFields': There are no fields inside this type."`;

exports[`Type Throws when trying to compile type without @InterfaceType decorator 1`] = `"@InterfaceType 'Bar': Class is not registered. Make sure it's decorated with @ObjectType decorator"`;
